server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,195 @@
+        return this.serverChannel != null;
+    }
+
+    // package private for testing
+    boolean enabled() {
+        return PORT.get(environment.settings()) != -1;
+    }
+
+    // package private for testing
+    ServerSocketChannel serverChannel() {
+        return serverChannel;
+    }
+
+    public BoundTransportAddress boundAddress() {
+        TransportAddress publishAddress = new TransportAddress(boundSocket.get());
+        return new BoundTransportAddress(new TransportAddress[] { publishAddress }, publishAddress);
+    }
+
+    ServerSocketChannel setupSocket() {
+        assert PORT.get(environment.settings()) >= 0;
+
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            InetAddress localhost = InetAddress.getLoopbackAddress();
+            int portNumber = PORT.get(environment.settings());
+
+            try {
+                InetSocketAddress socketAddress = new InetSocketAddress(localhost, portNumber);
+                serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                serverChannel.bind(socketAddress);
+
+                boundSocket.set(socketAddress);
+            } catch (Exception e) {
+                throw new BindTransportException("Failed to bind to " + NetworkAddress.format(localhost, portNumber), e);
+            }
+
+            return null;
+        });
+
+        return serverChannel;
+    }
+
+    @Override
+    protected void doStart() {
+        // Mark the service as active, we'll start the listener when ES is ready
+        this.active = enabled();
+        if (active == false) {
+            logger.debug("Readiness service is not enabled");
+        }
+    }
+
+    // package private for testing
+    synchronized void startListener() {
+        assert enabled();
+
+        if (this.serverChannel != null || this.active == false) {
+            return;
+        }
+
+        this.serverChannel = setupSocket();
+        this.listenerThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
+            assert serverChannel != null && listenerThreadLatch != null;
+            while (serverChannel.isOpen()) {
+                AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                    try (SocketChannel channel = serverChannel.accept()) {} catch (IOException e) {
+                        logger.debug("encountered exception while responding to readiness check request", e);
+                    } catch (Exception other) {
+                        logger.warn("encountered unknown exception while responding to readiness check request", other);
+                    }
+                    return null;
+                });
+            }
+            listenerThreadLatch.countDown();
+        }, "elasticsearch[readiness-service]").start();
+
+        logger.info("readiness service up and running on {}", boundAddress().publishAddress());
+
+        if (WRITE_PORTS_FILE_SETTING.get(environment.settings())) {
```

Comment:
This setting is really meant to be local to Node. Since it is really just for tests, can we set it in Node? We don't have integration tests that would read it after it is set and the node goes unready right?

Reply:
I originally moved this here when we had the port range scanning code, which would pick an unused port from the range 9400-9500. We don't bind the port until Elasticsearch is ready, so I don't know the port immediately when the service starts. This is no longer a problem when we hardcode the port, except for the ephemeral port. Let me see if I can wait until the service is ready, I think that should do it.